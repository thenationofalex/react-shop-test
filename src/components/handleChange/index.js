export default function handleChange (e) {
  const target = e.target
  const value = target.value
  const name = target.id
  console.log(name, value)
  this.setState({
    [name]: value.toString()
  })
}
