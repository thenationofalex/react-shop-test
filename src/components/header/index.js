import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

import { StyledHeader } from './styles'

class Header extends PureComponent {
  render () {
    return (
      <StyledHeader>
        <h1>{ this.props.title }</h1>
        { this.props.children }
      </StyledHeader>
    )
  }
}

Header.propTypes = {
  title: PropTypes.string,
  children: PropTypes.any
}

export default Header
