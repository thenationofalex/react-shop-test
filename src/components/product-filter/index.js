import React, { Fragment, PureComponent } from 'react'
import PropTypes from 'prop-types'
import flattenDeep from 'lodash/flattenDeep'
import uniq from 'lodash/uniq'

import { StyledSelect } from './styles'

let filterOptions = []

class ProductFilter extends PureComponent {
  render () {
    const { options, onChange } = this.props
    options.map(p => filterOptions.push(p.size))
    filterOptions = uniq(flattenDeep(filterOptions))
    return (
      <Fragment>
        <StyledSelect id='size' onChange={onChange}>
          <option value='reset'>Filter by size</option>
          {
            filterOptions.map(o => <option key={o} value={o}>{o}</option>)
          }
        </StyledSelect>
      </Fragment>
    )
  }
}

ProductFilter.propTypes = {
  options: PropTypes.array,
  onChange: PropTypes.func
}

export default ProductFilter
