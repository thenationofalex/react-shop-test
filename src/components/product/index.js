import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

import {
  StyledProduct,
  StyledProductInfo,
  StyledLabelBlock,
  StyledSale,
  StyledExclusive
} from './styles'

class Product extends PureComponent {
  render () {
    const { isSale, isExclusive, price, productImage, productName } = this.props
    return (
      <StyledProduct>
        <img src={`/static/${productImage}`} alt={productName} />
        <StyledLabelBlock>
          {
            isSale && (<StyledSale>sale</StyledSale>)
          }
          {
            isExclusive && (<StyledExclusive>exclusive</StyledExclusive>)
          }
        </StyledLabelBlock>
        <StyledProductInfo>
          <h2>{productName}</h2>
          <p>{price}</p>
        </StyledProductInfo>
      </StyledProduct>
    )
  }
}

Product.propTypes = {
  isSale: PropTypes.bool,
  isExclusive: PropTypes.bool,
  price: PropTypes.string,
  productImage: PropTypes.string,
  productName: PropTypes.string
}

export default Product
