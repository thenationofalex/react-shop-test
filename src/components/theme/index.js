export const theme = {
  colors: {
   green: '#009900',
   red: '#ff0000',
   blue: '#def1f4'
  }
}