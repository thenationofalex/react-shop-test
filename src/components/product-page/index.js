import React, { Fragment, PureComponent } from 'react'
import Head from 'next/head'
import PropTypes from 'prop-types'

import { StyledPageWrapper, StyledProductWrapper } from './styles'
import Header from '../header'
import ProductFilter from '../product-filter'
import Product from '../product'
import handleChange from '../handleChange'

class ProductPage extends PureComponent {
  handleChange = handleChange.bind(this)
  state = { productList: this.props.products }

  componentDidUpdate (prevProps, prevState) {
    if (prevState.size !== this.state.size) {
      let filteredItems = []
      if (this.state.size === 'reset') {
        filteredItems = this.props.products
      } else {
        this.props.products.forEach(product => {
          if (product.size.indexOf(this.state.size) !== -1) {
            filteredItems.push(product)
          }
        })
      }
      this.setState({ productList: filteredItems })
    }
  }
  render () {
    const { title } = this.props
    const { productList } = this.state
    return (
      <StyledPageWrapper>
        <Head>
          <title>{title}</title>
        </Head>
        <Fragment>
          <Header title={title}>
            <ProductFilter options={productList} onChange={this.handleChange} />
          </Header>
          <StyledProductWrapper>
            {
              productList.map(p => <Product key={p.index} {...p} />)
            }
          </StyledProductWrapper>
        </Fragment>
      </StyledPageWrapper>
    )
  }
}

ProductPage.propTypes = {
  title: PropTypes.string.isRequired,
  products: PropTypes.arrayOf(PropTypes.shape({
    index: PropTypes.number,
    isSale: PropTypes.bool,
    isExclusive: PropTypes.bool,
    price: PropTypes.string,
    productImage: PropTypes.string,
    productName: PropTypes.string,
    size: PropTypes.array
  }))
}

export default ProductPage
