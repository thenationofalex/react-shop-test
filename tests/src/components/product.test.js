import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import Product from '../../../src/components/product'
import {
  StyledLabel,
  StyledProduct,
  StyledLabelBlock,
  StyledSale,
  StyledExclusive,
  StyledProductInfo
} from '../../../src/components/product/styles'

import { theme } from '../../../src/components/theme'

const props = {
  theme,
  isSale: false,
  isExclusive: false,
  price: '20',
  productImage: 'foo.jpg',
  productName: 'Foo'
}

describe('Component - Product', () => {
  test('Renders', () => {
    const wrapper = shallow(<Product {...props} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  test('Renders - Sale', () => {
    const props = {
      theme,
      isSale: true,
      isExclusive: false,
      price: '20',
      productImage: 'foo.jpg',
      productName: 'Foo'
    }
    const wrapper = shallow(<Product {...props} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  test('Renders - Exclusive', () => {
    const props = {
      theme,
      isSale: false,
      isExclusive: true,
      price: '20',
      productImage: 'foo.jpg',
      productName: 'Foo'
    }
    const wrapper = shallow(<Product {...props} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  test('Styled Component - StyledLabel - Renders', () => {
    const wrapper = shallow(<StyledLabel {...props}>Some content</StyledLabel>)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  test('Styled Component - StyledProduct - Renders', () => {
    const wrapper = shallow(<StyledProduct {...props}>Some content</StyledProduct>)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  test('Styled Component - StyledLabelBlock - Renders', () => {
    const wrapper = shallow(<StyledLabelBlock {...props}>Some content</StyledLabelBlock>)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  test('Styled Component - StyledSale - Renders', () => {
    const wrapper = shallow(<StyledSale {...props}>Some content</StyledSale>)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  test('Styled Component - StyledExclusive - Renders', () => {
    const wrapper = shallow(<StyledExclusive {...props}>Some content</StyledExclusive>)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  test('Styled Component - StyledProductInfo - Renders', () => {
    const wrapper = shallow(<StyledProductInfo {...props}>Some content</StyledProductInfo>)
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
