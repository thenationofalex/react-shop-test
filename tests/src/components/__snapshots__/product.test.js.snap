// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - Product Renders - Exclusive 1`] = `
<Styled(div)>
  <img
    alt="Foo"
    src="/static/foo.jpg"
  />
  <Styled(div)>
    <Styled(div)>
      exclusive
    </Styled(div)>
  </Styled(div)>
  <Styled(div)>
    <h2>
      Foo
    </h2>
    <p>
      20
    </p>
  </Styled(div)>
</Styled(div)>
`;

exports[`Component - Product Renders - Sale 1`] = `
<Styled(div)>
  <img
    alt="Foo"
    src="/static/foo.jpg"
  />
  <Styled(div)>
    <Styled(div)>
      sale
    </Styled(div)>
  </Styled(div)>
  <Styled(div)>
    <h2>
      Foo
    </h2>
    <p>
      20
    </p>
  </Styled(div)>
</Styled(div)>
`;

exports[`Component - Product Renders 1`] = `
<Styled(div)>
  <img
    alt="Foo"
    src="/static/foo.jpg"
  />
  <Styled(div) />
  <Styled(div)>
    <h2>
      Foo
    </h2>
    <p>
      20
    </p>
  </Styled(div)>
</Styled(div)>
`;

exports[`Component - Product Styled Component - StyledExclusive - Renders 1`] = `
<div
  className="css-1xoizpk"
>
  Some content
</div>
`;

exports[`Component - Product Styled Component - StyledLabel - Renders 1`] = `
<div
  className="css-1l78bc"
>
  Some content
</div>
`;

exports[`Component - Product Styled Component - StyledLabelBlock - Renders 1`] = `
<div
  className="css-1e5wpvt"
>
  Some content
</div>
`;

exports[`Component - Product Styled Component - StyledProduct - Renders 1`] = `
<div
  className="css-1lk0485"
>
  Some content
</div>
`;

exports[`Component - Product Styled Component - StyledProductInfo - Renders 1`] = `
<div
  className="css-1v63bf1"
>
  Some content
</div>
`;

exports[`Component - Product Styled Component - StyledSale - Renders 1`] = `
<div
  className="css-xw1t6g"
>
  Some content
</div>
`;
